### PR TITLE
Implement delta and period

### PR DESCRIPTION
**Pull request by bubriks**
[#86](https://github.com/Wappsto/wappsto-python/issues/86)

### CODE REVIEW

**Comment by bubriks** [comment]
Needs better tests and to improve code organization

File: test/wappsto_test.py [hunk 1]
```diff
@@ -428,10 +434,22 @@ def test_send_value_update_number_type(self, input, step_size, expected):
         value = device.values[0]
         value.data_type == "number"
         value.number_step = step_size
+        if delta:
+            value.last_update_of_control = 0
```
Comment (AndreasBomholtz) [inline test/wappsto_test.py]: Why is it called control when it is a report?

File: test/wappsto_test.py [hunk 2]
```diff
@@ -428,10 +434,22 @@ def test_send_value_update_number_type(self, input, step_size, expected):
...
+            value.set_delta(delta)
+            if not abs(input - value.last_update_of_control) >= value.delta:
```
Comment (AndreasBomholtz) [inline test/wappsto_test.py]: Why do you need the not? Can't you not just change >= to <?

File: wappsto/connection/communication.py [hunk 3]
```diff
@@ -856,6 +853,15 @@ def close(self):
         Closes the socket object connection.
         """
         self.wapp_log.info("Closing connection...")
+
+        for device in self.network.devices:
+            for value in device.values:
+                if value.timer.is_alive():
+                    msg = "Value: {} is no longer periodically sending updates."
+                    msg = msg.format(value.uuid)
+                    self.wapp_log.info(msg)
```
Comment (AndreasBomholtz) [inline wappsto/connection/communication.py]: This should be a debug message.

File: wappsto/connection/network_classes/value.py [hunk 4]
```diff
@@ -88,16 +84,14 @@ def __init__(
         self.string_max = string_max
         self.blob_encoding = blob_encoding
         self.blob_max = blob_max
-        self.period = period
-        self.delta = delta
+        self.period = None
```
Comment (AndreasBomholtz) [inline wappsto/connection/network_classes/value.py]: Will you read the period and delta from the json file later?

File: wappsto/connection/network_classes/value.py [hunk 5]
```diff
@@ -121,22 +115,35 @@ def set_period(self, period):
...
         """
         try:
-            if self.get_report_state() is not None and int(period) > 0:
+            period = int(period)
+            if self.get_report_state() is not None and period > 0:
                 self.period = period
-
+                self.__set_timer()
             else:
                 self.wapp_log.warning("Cannot set the period for this value.")
```
Comment (AndreasBomholtz) [inline wappsto/connection/network_classes/value.py]: This warning is wrong if the period is 0.

File: wappsto/connection/network_classes/value.py [hunk 6]
```diff
@@ -149,9 +156,10 @@ def set_delta(self, delta):
...
         """
         try:
+            delta = float(delta)
             if (self.__is_number_type()
                     and self.get_report_state()
-                    and float(delta) > 0):
+                    and delta > 0):
                 self.delta = delta
             else:
                 self.wapp_log.warning("Cannot set the delta for this value.")
```
Comment (AndreasBomholtz) [inline wappsto/connection/network_classes/value.py]: This warning is wrong if delta is zero.

File: wappsto/connection/network_classes/value.py [hunk 7]
```diff
@@ -471,13 +394,66 @@ def update(self, data_value, timestamp=None):
...
+        Args:
+            return_value: default return value.
+
+        Returns:
+            True/False indicating the result of operation.
+
+        """
+        if self.period is not None:
+            # period should work
+            if threading.current_thread() == self.timer or not self.timer.is_alive():
```
Comment (AndreasBomholtz) [inline wappsto/connection/network_classes/value.py]: Is the main.py sending updates in the timer thread or how do you expect this check to work?

File: test/wappsto_test.py [hunk 8]
```diff
@@ -475,10 +498,21 @@ def test_send_value_update_text_type(self, input, max, expected, type):
         value.data_type = type
         value.string_max = max
         value.blob_max = max
+        if delta:
+            value.last_update_of_control = 0
```
Comment (AndreasBomholtz) [inline test/wappsto_test.py]: Should this not be report?

File: wappsto/connection/network_classes/value.py [hunk 9]
```diff
@@ -88,16 +86,17 @@ def __init__(
...
-        self.period = period
-        self.delta = delta
         self.report_state = None
         self.control_state = None
         self.callback = self.__callback_not_set
-        self.last_update_of_report = self.get_now()
-        self.reporting_thread = None
+
+        self.timer = threading.Timer(None, None)
         self.last_update_of_control = None
```
Comment (AndreasBomholtz) [inline wappsto/connection/network_classes/value.py]: Should this not be report?

File: wappsto/connection/network_classes/value.py [hunk 10]
```diff
@@ -471,13 +410,66 @@ def update(self, data_value, timestamp=None):
...
+            return_value: default return value.
+
+        Returns:
+            True/False indicating the result of operation.
+
+        """
+        if self.period is not None:
+            # period should work
+            if self.timer_elapsed:
+                # this is timer thread or timer is not running
```
Comment (AndreasBomholtz) [inline wappsto/connection/network_classes/value.py]: Wrong comment

**Pull request merged**